sdks/java/core/src/main/java/org/apache/beam/sdk/coders/SerializableCoder.java

```diff
@@ -57,14 +63,43 @@
     return new SerializableCoder<>(clazz, type);
   }
 
+  @Override
+  public boolean consistentWithEquals() {
+    return false;
+  }
+
+  /**
+   * The structural value of the object is the object itself. The {@link SerializableCoder} should
+   * be only used for objects with a proper {@link Object#equals} implementation.
+   */
+  @Override
+  public Object structuralValue(T value) {
+    return value;
+  }
+
   /**
    * Returns a {@link SerializableCoder} instance for the provided element class.
    * @param <T> the element type
    */
   public static <T extends Serializable> SerializableCoder<T> of(Class<T> clazz) {
+    checkEqualsMethodDefined(clazz);
     return new SerializableCoder<>(clazz, TypeDescriptor.of(clazz));
   }
 
+  private static <T extends Serializable> void checkEqualsMethodDefined(Class<T> clazz) {
+    try {
+      Method method = clazz.getMethod("equals", Object.class);
+      if (clazz.isInterface() || !clazz.equals(method.getDeclaringClass())) {
+        LOG.warn("Can't verify serialized elements of type {} have well defined equals method. "
+                + "This may produce incorrect results on some {}", clazz.getSimpleName(),
+            PipelineRunner.class.getSimpleName());
+      }
+    } catch (NoSuchMethodException e) {
```

Comment:
This is actually possible now, because you get the method before you check to see if `clazz` is an interface

Reply:
Sorry about that, added more testing so this verified.